core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -450,56 +457,74 @@ private[kafka] class Processor(val id: Int,
 
   override def run() {
     startupComplete()
-    while (isRunning) {
-      try {
-        // setup any new connections that have been queued up
-        configureNewConnections()
-        // register any new responses for writing
-        processNewResponses()
-        poll()
-        processCompletedReceives()
-        processCompletedSends()
-        processDisconnected()
-      } catch {
-        // We catch all the throwables here to prevent the processor thread from exiting. We do this because
-        // letting a processor exit might cause a bigger impact on the broker. Usually the exceptions thrown would
-        // be either associated with a specific socket channel or a bad request. We just ignore the bad socket channel
-        // or request. This behavior might need to be reviewed if we see an exception that need the entire broker to stop.
-        case e: ControlThrowable => throw e
-        case e: Throwable =>
-          error("Processor got uncaught exception.", e)
+    try {
+      while (isRunning) {
+        try {
+          // setup any new connections that have been queued up
+          configureNewConnections()
+          // register any new responses for writing
+          processNewResponses()
+          poll()
+          processCompletedReceives()
+          processCompletedSends()
+          processDisconnected()
+        } catch {
+          // We catch all the throwables here to prevent the processor thread from exiting. We do this because
+          // letting a processor exit might cause a bigger impact on the broker. This behavior might need to be
+          // reviewed if we see an exception that needs the entire broker to stop. Usually the exceptions thrown would
+          // be either associated with a specific socket channel or a bad request. These exceptions are caught and processed
+          // by the individual methods above which close the failing channel and continue processing other channels.
+          // So this catch block should only ever see ControlThrowables.
+          case e: Throwable => processException("Processor got uncaught exception.", e)
+        }
       }
+    } finally {
+      debug("Closing selector - processor " + id)
+      swallowError(closeAll())
+      shutdownComplete()
+    }
+  }
+
+  private def processException(errorMessage: String, throwable: Throwable) {
+    throwable match {
+      case e: ControlThrowable => throw e
+      case e => error(errorMessage, e)
     }
+  }
 
-    debug("Closing selector - processor " + id)
-    swallowError(closeAll())
-    shutdownComplete()
+  private def processChannelException(channelId: String, errorMessage: String, throwable: Throwable) {
+    openOrClosingChannel(channelId).foreach(_ => {
+      error(s"Closing socket for ${channelId} because of error", throwable)
+      close(selector, channelId)
+      inflightResponses.remove(channelId).foreach(response => updateRequestMetrics(response.request))
```

Comment:
Maybe we should have a `close(channel)` method in `Processor` that does both of the above. There are a couple of other cases where we are calling the `close` method in `AbstractServerThread`. Sorry for taking a few iterations, I am trying to make things as consistent as possible so that we don't miss things.

Reply:
@ijuma Thank you for the reviews. I couldn't find a good reason for the `close(selector, connectionId` method to live in `AbstractServerThread`, so I took the liberty of moving it to `Processor`. Connection quotas are updated in `Processor` for disconnected anyway. Let me know if I missed something.

Reply:
Yeah, that's better. I was going to suggest that, but when looking quickly thought there was one usage that needed it to be in `AbstractServerThread`, but that's not correct as you've shown.